compiler_opt/rl/regalloc/config.py

```diff
@@ -54,6 +55,8 @@ def get_regalloc_signature_spec():
                        'nr_broken_hints', 'nr_urgent', 'nr_rematerializable')))
   observation_spec['progress'] = tensor_spec.BoundedTensorSpec(
       dtype=tf.float32, shape=(), name='progress', minimum=0, maximum=1)
+  observation_spec['lr_use_def_instructions'] = tensor_spec.BoundedTensorSpec(
+      dtype=tf.int64, shape=(33), name='lr_use_def_instructions', minimum=0, maximum=17716)
```

Comment:
I assume this is the length of the sequence of instructions? 33 comes from the number of 32 live ranges in physical register + 1 candidate live range. technically there is no correlation between these two numbers

Reply:
the most naive way to add the feature is logging down all the instructions, so it is of various length, may be 10, may be 100, etc. then you get another feature of dimension 33, that each of them describes which instructions this live range occupies, so it can be 33 * 100 (assume 100 is the number of instructions) 1/0 binary features or it can be 33 * (various length of list), the latter is the sparse representation of the 100 dimension 1/0 binary feature.

In the code here, first get the 100 * 16 dimension embedding, then use the binary feature to add all the live-range occupied instructions embedding together

Reply:
> I assume this is the length of the sequence of instructions? 33 comes from the number of 32 live ranges in physical register + 1 candidate live range. technically there is no correlation between these two numbers

Yes, this is just the length of the maximum number of instructions that can be passed to the model. I was originally planning on just encoding 10 use/def instructions because that seems to be enough to encode all the use/def instructions for the majority of live ranges, but because the preprocessing layer is set to concatenate, this tensor has to have a length of 33 or otherwise the concatenate layer won't work. The rest of the tensor is just zero-padded. Like you mentioned, there is no correlation between these two numbers, just an artificial limitation of the preprocessing combining layer.

Reply:
> the most naive way to add the feature is logging down all the instructions, so it is of various length, may be 10, may be 100, etc. then you get another feature of dimension 33, that each of them describes which instructions this live range occupies, so it can be 33 * 100 (assume 100 is the number of instructions) 1/0 binary features or it can be 33 * (various length of list), the latter is the sparse representation of the 100 dimension 1/0 binary feature.
> 
> In the code here, first get the 100 * 16 dimension embedding, then use the binary feature to add all the live-range occupied instructions embedding together

I was planning a somewhat similar implementation. Instead of logging down all the instructions sequentially like you mentioned, I was planning on extracting instructions per live range  (experimenting with just use/def and all instructions)(pretty simple in the LLVM API, getting all of the instructions sequentially might be possible, but I'm not currently aware of such facilities to do so currently) into a 2d array (33, number of desired instructions - probably 10 right now), running this through an embedding layer, and then reshaping to (33, embedding dimensions * number of desired instructions) so that it will sail smoothly through the concatenate layer. This should simplify transferring features since there is only one, and then fit in more nicely (at least from current knowledge) with the LLVM APIs.

Reply:
the problem of this approach is that it is duplicating the information. If an instruction is use/def'ed by X live ranges, then the information is duplicated X times in logging. This is not very efficient

With the logging of the sequence of instructions and the usage information, you can also easily get what you want by doing a multiplication between these two.

Reply:
That's a good point. I'll look further into the LLVM APIs/talk with Mircea about options to obtain this specific data. Thanks for the outline/discussion.